### ISSUE TITLE
"Unsafe call to import for argument 0" warning in content script is unfixable

### ISSUE THREAD

**Issue by AprilSylph** [mozilla/addons-linter#3846]
### Describe the problem and steps to reproduce it:

Using ES modules in a content script is impossible to do without producing an `UNSAFE_VAR_ASSIGNMENT` warning.

Reproduction steps:
1. Try to use dynamic `import` within a content script to import an ES module, with a path relative to the content script. This obviously fails since the path is resolved relative to the webpage, while the module is in the addon files; the Firefox console gives the warning "WebExtension content scripts may only load modules with moz-extension URLs".
2. Try to resolve the issue by instead passing a path relative to the addon manifest to `browser.runtime.getURL`, and use the result in `import`. This successfully imports the module, but causes addons-linter to produce a warning message ("Unsafe call to import for argument 0") which cannot be fixed.

### What happened?

```json
{
    "_type": "warning",
    "code": "UNSAFE_VAR_ASSIGNMENT",
    "message": "Unsafe call to import for argument 0",
    "description": "Due to both security and performance concerns, this may not be set using dynamic values which have not been adequately sanitized. This can lead to security issues or fairly serious performance degradation.",
    "column": 65,
    "file": "content_scripts/main.js",
    "line": 19
}
```

### What did you expect to happen?

Either no warning (since dynamic `import` in addons is not allowed to import non-bundled modules), or a way to improve my code to resolve the issue the warning reports.

I tried installing `eslint-plugin-no-unsanitized` and following the ["Fixing Rule Violations"](https://github.com/mozilla/eslint-plugin-no-unsanitized/blob/master/docs/rules/fixing-violations.md) doc to eliminate the errors the plugin produced by sanitizing the constructed `moz-extension` URLs, only to find that addons-linter explicitly disables recognising escapers:

https://github.com/mozilla/addons-linter/blob/0cccb218879e7c07663b4b2eb8242603758e2f39/src/const.js#L16

### Anything else we should know?

The addon file that produced the quoted warning is [here](https://github.com/AprilSylph/XKit-Rewritten/blob/master/src/content_scripts/main.js). I do use template strings in this file, but restructuring this content script to only use static strings would still produce the warning, since anything that `browser.runtime.getURL` returns is apparently unsafe.

My apologies if this is hard to follow. Let me know if there's anything I can do to improve my explanation.

**Comment by rpl** [comment]
@mozfreddyb it looks that eslint-plugin-no-unsanitized is reporting a "Unsafe call to import for argument 0" linting warnings on each dynamic import expressions that is using a non-literal parameter, is this something that has been reported and we are already aware of?

I'll dig into it more asap, but in the meantime I wanted to hear your perspective about it.

**Comment by rpl** [comment]
Well, I think I can actually answer my own question:
- this seems to be the expected behavior based on the test cases added in [the pull request mozilla/eslint-plugin-no-unsanitized#150](https://github.com/mozilla/eslint-plugin-no-unsanitized/pull/150/files#diff-0d3319441e61cd858549d3fe5a8db795e4e3ad4ac98193e94a8fdc87bb8cbfaf)

**Comment by mozfreddyb** [comment]
Yeah, I see how this is an annoying issue for the AMO world. This eslint plugin was not designed for this use-case.
The risks are along the lines of `import(attackerControlled)` and completely dynamic scripts with `data:` URLs.


So here are some options to consider:
- [disable the "import" rule in the plugin config](https://github.com/mozilla/eslint-plugin-no-unsanitized/blob/master/docs/rules/customization.md) and enforce that extensions are self-contained through other means (I know there are *some* checks in Gecko?)
- configure the plugin to consider the function browser.runtime.getURL to be a valid escaper (need to write some tests to ensure that custom functions of the same name can't confuse you)

Let me know if that helps!

**Comment by AprilSylph** [comment]
> completely dynamic scripts with `data:` URLs

are disallowed in content scripts as per [Bug 1587336](https://bugzilla.mozilla.org/show_bug.cgi?id=1587336#c8), but I don't know about background scripts. I suppose a malicious WebExtension could theoretically still fool a different WebExtension into loading the former's modules though, since they would both be on the allowed `moz-extension` scheme.

Considering `browser.runtime.getURL` to be a valid escaper would solve my specific problem, assuming it isn't ignored by addons-linter like `Sanitizer.escapeHTML` is.

**Comment by mozfreddyb** [comment]
The addons-linter configured the eslint-plugin-no-unsanitized to allow Sanitizer.escapeHMTL? I know this is the default config, but I suppose that should be disabled.. :confused:

**Comment by AprilSylph** [comment]
No, addons-linter is configured to disallow `Sanitizer.escapeHTML`. That's why I'm saying `browser.runtime.getURL` would have to get different treatment to suppress the warnings I'm getting.

**Comment by naser7717** [comment]
ع

**Comment by rpl** [comment]
Thanks for reporting this issue,
we discussed about this issue during our last triage meeting and agreed to close it as wontfix, with the following rationale:

- the related addons-linter rule is categorized as a warning (on purpose) and so it doesn't block submitting the addon xpi to AMO for being signed
- we can't be 100% sure based on the static analysis of the AST and of the extension manifest file if `browser.runtime.getURL` is the legit extension API provided by Firefox (we can't be even 100% sure if a certain source file is only executed in extension contexts or also injected into a webpage) and in the context of the addons-linter we can't configure it as an allowed sanitizer function.
- we prefer to leave the warning unmodified, it is meant to be an hint for the addon reviewers for code that may be worth double-checking

**Comment by enchanted-sword** [comment]
FYI: This was fixed in https://github.com/mozilla/addons-linter/pull/4739, `browser.runtime.getURL` and `chrome.runtime.getURL` are now considered valid escapes.
